tmt/base.py

```diff
@@ -1001,9 +1001,20 @@ def __init__(
     def _expand_node_data(self, data: T) -> T:
         """ Recursively expand variables in node data """
         if isinstance(data, str):
-            # This cast is tricky: we get a string, and we return a string, so T -> T hold,
-            # yet mypy does not recognize this, and we need to help with an explicit cast().
-            return cast(T, os.path.expandvars(data))
+            # expand env variables
+            data_expanded_vars = os.path.expandvars(data)
+            # convert context values to string and define env variables using context
+            # keys and values
+            dict_context_vars = {
+                key: ','.join(value)
+                for (key, value) in self._fmf_context().items()
+                }
+            # convert $@{} to ${} and expand context using temporarily defined
+            # environment variables
+            with tmt.utils.modify_environ(dict_context_vars):
```

Comment:
Using the `modify_environment` from within the `_expand_node_data()` method means it can be called many times when more complex dictionaries are process, probably not a big deal, but perhaps it might make sense to handle this simlarly as for `environment`: Update the variables once and do the whole expansion including recursion under it.

Reply:
Updated to optional function argument per @happz 's request.